### PR TITLE
fix(Core) Quest Relations and Gossip Quest Icons WIP

### PR DESCRIPTION
**Pull request by malcrom**
After more research I'm starting to understand quest relations better. There are different ways that auto complete and repeatable quest are handled. I think what I have done so far here makes all quest relations behave blizzlike.

In regards to the gossip quest icons I have dailies now showing as blue but am still having trouble making the auto complete quest show as a yellow question make.

## Changes Proposed:
-  Changed quest relations handling in GetQuestDialogStatus.
-  Add check to SendQuestGiverQuestList for dailies.

## Issues Addressed:
- Improper relations for many questgivers.
- Improper icon for daily quest.

## Tests Performed:
- Builds
- Tested on multiple npcs with different quest options. 

Auto complete by flag non repeatable. Proper Relations.
Allison

Auto complete by method non repeatable. Proper Relations. Still issue with icons
Clavicus

Auto complete by method repeatable. Proper Relations. Still issue with icons
Mathredis

Auto complete by method repeatable quest level -1. Proper Relations
tiza

Repeatable daily. Proper Relations. Proper icons
daily

## Known Issues and TODO List:

Mathredis gossip should look like this. Clavicus has same issue. Auto complete quest should have yellow question marks. 

Freemanpng

Here the daily is complete but the normal quest is not taken. I'm not sure if this is correct behavior or if there should be a blue question mark over his head? 
daily complete

Code in use seems to be working properly. There could be possible cleanup. Any help with the yellow question marks would be appreciated.

### CODE REVIEW

File: src/server/game/Entities/Player/PlayerQuest.cpp [hunk 1]
```diff
@@ -69,9 +69,11 @@ void Player::PrepareQuestMenu(ObjectGuid guid)
     for (QuestRelations::const_iterator i = objectQIR.first; i != objectQIR.second; ++i)
     {
         uint32 quest_id = i->second;
+        Quest const* quest = sObjectMgr->GetQuestTemplate(quest_id);
         QuestStatus status = GetQuestStatus(quest_id);
+
         if (status == QUEST_STATUS_COMPLETE)
-            qm.AddMenuItem(quest_id, 4);
+            qm.AddMenuItem(quest_id, 4);
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Player/PlayerQuest.cpp:76]: revert this please

File: src/server/game/Entities/Player/PlayerQuest.cpp [hunk 2]
```diff
@@ -1656,30 +1663,37 @@ QuestGiverStatus Player::GetQuestDialogStatus(Object* questgiver)
...
-                    {
-                        if (quest->IsDaily())
+                        if (quest->IsRepeatable())
                         {
                             if (isLowLevel)
-                                result2 = DIALOG_STATUS_LOW_LEVEL_AVAILABLE_REP;
+                                result2 = DIALOG_STATUS_LOW_LEVEL_REWARD_REP;
                             else
-                                result2 = DIALOG_STATUS_AVAILABLE_REP;
+                                result2 = DIALOG_STATUS_REWARD_REP;
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Player/PlayerQuest.cpp:1674]: what about:

```cpp
result2 = isLowLevel ? DIALOG_STATUS_LOW_LEVEL_REWARD_REP : DIALOG_STATUS_REWARD_REP
```

it's the same thing as your `if/else` but with 1 line of code

File: src/server/game/Entities/Player/PlayerQuest.cpp [hunk 3]
```diff
@@ -1656,30 +1663,37 @@ QuestGiverStatus Player::GetQuestDialogStatus(Object* questgiver)
...
+                            result2 = DIALOG_STATUS_REWARD;
                         }
                     }
+                    // If quest has daily flag
+                    else if (quest->IsDaily())
+                    {
+                        if (isLowLevel)
+                            result2 = DIALOG_STATUS_LOW_LEVEL_AVAILABLE_REP;
+                        else
+                            result2 = DIALOG_STATUS_AVAILABLE_REP;
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Player/PlayerQuest.cpp:1687]:
```suggestion
                        result2 = isLowLevel ? DIALOG_STATUS_LOW_LEVEL_AVAILABLE_REP : DIALOG_STATUS_AVAILABLE_REP;
```

File: src/server/game/Entities/Player/PlayerQuest.cpp [hunk 4]
```diff
@@ -1656,30 +1663,37 @@ QuestGiverStatus Player::GetQuestDialogStatus(Object* questgiver)
...
+                    }
+                    // If quest is normal type
+                    else
+                    {
+                        if (isLowLevel)
+                            result2 = DIALOG_STATUS_LOW_LEVEL_AVAILABLE;
+                        else
+                            result2 = DIALOG_STATUS_AVAILABLE;
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Player/PlayerQuest.cpp:1695]:
```suggestion
                        result2 = isLowLevel ? DIALOG_STATUS_LOW_LEVEL_AVAILABLE : DIALOG_STATUS_AVAILABLE;
```

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by syssneck** [comment]
Hey, I tested it, but I'm having some issues with it.
I can't see some quests, that are auto complete by method and repeatable. I don't see any quests for Mathredis for example, are there any prerequisites that I missed?
grafik

Also, this doesn't look right. Both quests are done, but I still see a yellow exclamation mark:
grafik

**Comment by malcrom** [comment]
Mathredis has been properly scripted now. You will not see a quest unless you have a book on you. So that is not an issue. The second picture  looks odd. I will have to look into this. Thanks for testing.

**Comment by claudiodfc** [comment]
Does this need to be re-tested or is it WIP?

**Comment by malcrom** [comment]
> Does this need to be re-tested or is it WIP?

Still WIP

**Pull request closed without merging**